cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +122,167 @@ public Void call() throws Exception {
     }
   }
 
+  /**
+   * Clean log files which does not have corresponding meta data
+   * @param tillTime time till the meta data will be deleted.
+   * @param namespacedLogBaseDirMap namespace to directory map
+   * @param parentDirs parent directories for deleted files
+   */
+  @VisibleForTesting
+  void cleanFilesWithoutMeta(final long tillTime, final Map<String, NamespaceId> namespacedLogBaseDirMap,
+                             final SetMultimap<String, Location> parentDirs) throws Exception {
+    LOG.info("Starting deletion of log files older than {} without metadata", tillTime);
+    List<NamespaceMeta> namespaces = namespaceQueryAdmin.list();
+    // For all the namespaces present in NamespaceMeta, gather log files from disk and log meta. Then delete the log
+    // files for which metadata is not present
+    for (NamespaceMeta namespaceMeta : namespaces) {
+      final NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      final Location namespacedLogBaseDir = LoggingContextHelper.getNamespacedBaseDirLocation(namespacedLocationFactory,
+                                                                                              logBaseDir, namespaceId,
+                                                                                              impersonator);
+      // processor to get list of log files older than retention duration
+      final Processor<LocationStatus, Set<Location>> diskFilesProcessor = getDiskFilesProcessor(namespaceId, tillTime);
+      // gather all the disk log files older than tillTime for a given namespace
+      if (namespacedLogBaseDir.exists()) {
+        try {
+          impersonator.doAs(namespaceId, new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+              Locations.processLocations(namespacedLogBaseDir, true, diskFilesProcessor);
+              return null;
+            }
+          });
+        } catch (Exception e) {
+          LOG.warn("Got exception while accessing path {}", namespacedLogBaseDir, e);
+        }
+        Set<Location> diskFileLocations = diskFilesProcessor.getResult();
+
+        // get the logging context for a given namespace
+        LoggingContext loggingContext = new LogNamespaceLoggingContext(namespaceId.getNamespace());
+
+        // get all the log meta files for a given namespace
+        FileMetaDataManager.TableKey nextTableKey = new FileMetaDataManager.TableKey(loggingContext.getLogPartition(),
+                                                                                     null);
+        Processor<URI, Set<URI>> metaProcessor = getMetaFilesProcessor();
+        // Get all the log metadata in batches of 100
+        do {
+          nextTableKey = fileMetaDataManager.scanFiles(nextTableKey, 100, metaProcessor);
```

Comment:
@poornachandra If I am understanding it correctly, in above algorithm you are assuming that we will have log files without meta in 50k chunk. Multiple runs of `cleanFilesWithoutMeta()` would not guarantee scanning of all the disk files in given namespace. For example, lets say we have 60k log files present in a given namespace. Out of which 0-55k log files are less than `tillTime`. The log file without meta may be present out side 50k chunk we scanned in this run.  In the next run also we scan 50k files. However, remaining 5k files from first run might still not get processed and those 5k files will still be on the disk and that chunk might have files without metadata. Not sure how can we guarantee that those 5k log files are also processed? While scanning disk files, we always provide `namespacedLogDir` as start location and do not store any information about where previous run ended.